providers/powerdns/powerdnsProvider.go

```diff
@@ -0,0 +1,269 @@
+				},
+			})
+		} else {
+			ttl := desiredRecords[label][0].TTL
+			records := []zones.Record{}
+			for _, recordContent := range desiredRecords[label] {
+				records = append(records, zones.Record{
+					Content: recordContent.GetTargetCombined(),
+				})
+			}
+			corrections = append(corrections, &models.Correction{
+				Msg: concatMessage(msgs),
+				F: func() error {
+					return api.client.Zones().AddRecordSetToZone(context.Background(), api.ServerName, dc.Name, zones.ResourceRecordSet{
+						Name:    labelName,
+						Type:    labelType,
+						TTL:     int(ttl),
+						Records: records,
+					})
+				},
+			})
+		}
+	}
+
+	// DNSSec corrections
+	// TODO: Implementing of on-demand DNSSec creation in library
+
+	return corrections, nil
+}
+
+// EnsureDomainExists adds a domain to the DNS service if it does not exist
+func (api *PowerDNS) EnsureDomainExists(domain string) error {
+	if zone, _ := api.client.Zones().GetZone(context.Background(), api.ServerName, domain); zone.ID != "" {
+		return nil
+	}
+
+	_, err := api.client.Zones().CreateZone(context.Background(), api.ServerName, zones.Zone{
+		Name:        domain,
+		Type:        zones.ZoneTypeZone,
+		DNSSec:      api.DNSSecOnCreate,
+		Nameservers: api.DefaultNS,
+	})
+	return err
+}
+
+// toRecordConfig converts a PowerDNS DNSRecord to a RecordConfig. #rtype_variations
+func toRecordConfig(domain string, r zones.Record, ttl int, name string, rtype string) (*models.RecordConfig, error) {
+	// trimming trailing dot and domain from name
+	if strings.HasSuffix(name, domain+".") {
+		name = strings.TrimSuffix(name, domain+".")
+	}
+	if strings.HasSuffix(name, ".") {
+		name = strings.TrimSuffix(name, ".")
+	}
+
+	rc := &models.RecordConfig{
+		TTL:      uint32(ttl),
+		Original: r,
+		Type:     rtype,
+	}
+	rc.SetLabel(name, domain)
+
+	content := r.Content
+	switch rtype {
+	case "CNAME", "NS":
+		return rc, rc.SetTarget(dnsutil.AddOrigin(ensureDotSuffix(content), domain))
+	case "CAA":
+		return rc, rc.SetTargetCAAString(content)
+	case "MX":
+		return rc, rc.SetTargetMXString(ensureDotSuffix(content))
+	case "SRV":
+		return rc, rc.SetTargetSRVString(content)
+	case "TXT":
+		// Remove quotes if it is a TXT record.
+		if !strings.HasPrefix(content, `"`) || !strings.HasSuffix(content, `"`) {
+			return nil, errors.New("Unexpected lack of quotes in TXT record from PowerDNS")
+		}
+		return rc, rc.SetTargetTXT(content[1 : len(content)-1])
+	default:
+		return rc, rc.PopulateFromString(rtype, content, domain)
+	}
+}
+
+// concatMessage creates a single string of messages of a slice of messages
+func concatMessage(messages []string) string {
+	var msg string
+	for i, singleMsg := range messages {
+		if len(messages) > 1 && i != len(messages)-1 {
+			msg += singleMsg + "\n    "
+		} else {
+			msg += singleMsg
+		}
+	}
+	return msg
+}
+
+// ensureDotSuffix ensures that the trailing dot is there
```

Comment:
We might no longer need this function.

When we first wrote dnscontrol there was a lot of places where we stripped or added a `.` "just in case".  In all those places, it turned out we were adding complexity that wasn't needed.  The code is now very careful (for example) to make sure that:

* label.NameFQDN never ends in a "."
* label.GetTarget() always ends with a "." (unless it is an IP address, etc.)

That should simplify your code.